src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,296 @@
+
+            var root = (CompilationUnitSyntax) await _syntaxTree.GetRootAsync();
+
+            _syntaxWalker = new CSharpAlgoValidationWalker(_sourceText);
+            _syntaxWalker.Visit(root);
+
+            validationMessages.AddRange(_syntaxWalker.GetMessages());
+
+            if (ErrorExists(validationMessages))
+                return CreateAndSetValidationResult(out _syntaxValidationResult, false, validationMessages);
+
+            // Semantic validation
+
+            var coreLib = MetadataReference.CreateFromFile(typeof(object).Assembly.Location);
+            var fxLibs = AllowedFxLibs
+                .Select(l => MetadataReference.CreateFromFile(
+                    Assembly.Load(
+                        $"{l.Item1}, Version={l.Item2}, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a").Location))
+                .ToArray();
+
+            _compilation = CSharpCompilation.Create("CodeValidation")
+                .AddSyntaxTrees(_syntaxTree)
+                .AddReferences(coreLib)
+                .AddReferences(fxLibs)
+                .AddReferences(await NuGetReferenceProvider.GetReferences())
+                .WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            _semanticModel = _compilation.GetSemanticModel(_syntaxTree, false);
+            var semanticDiagnostics = _semanticModel.GetDiagnostics();
+
+            validationMessages.AddRange(semanticDiagnostics.Select(DiagnosticToValidationMessage));
+
+            return CreateAndSetValidationResult(out _syntaxValidationResult,
+                !ErrorExists(validationMessages),
+                validationMessages);
+        }
+
+        public Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var algoClassFullName = _syntaxWalker.NamespaceNode == null
+                ? $"{_syntaxWalker.ClassNode.Identifier}"
+                : $"{_syntaxWalker.NamespaceNode.Name}.{_syntaxWalker.ClassNode.Identifier}";
+
+            var newAssembly = GenerateAssembly();
+            var newType = newAssembly.GetType(algoClassFullName);
+            var metadata = ExtractMetadata(newType);
+
+            return Task.FromResult(metadata);
+        }
+
+        private AlgoMetaDataInformation ExtractMetadata(Type algoType)
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            //Get public properties that have public setter
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                .Where(x => x.GetSetMethod() != null);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                //Base parameters
+                if (algoProperty.PropertyType.BaseType != typeof(AbstractFunction)
```

Comment:
Redundant check, `typeof().IsAssignableFrom` also works for indirect inheritance